foundations/src/telemetry/tokio/mod.rs

```diff
@@ -0,0 +1,364 @@
+//! | tokio_runtime_worker_busy_duration_micros_total  | [`tokio::runtime::RuntimeMetrics::worker_total_busy_duration`]      | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_local_schedules_total       | [`tokio::runtime::RuntimeMetrics::worker_local_schedule_count`]     | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_overflows_total             | [`tokio::runtime::RuntimeMetrics::worker_overflow_count`]           | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_local_queue_depth           | [`tokio::runtime::RuntimeMetrics::worker_local_queue_depth`]        | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_mean_poll_time_micros       | [`tokio::runtime::RuntimeMetrics::worker_mean_poll_time`]           | runtime_name?, runtime_id?, worker_idx |
+//!
+//! ## Example
+//! ```no_run
+//! # use std::thread;
+//! # use std::time::Duration;
+//! # use foundations::telemetry::tokio::RuntimeMonitor;
+//! // create and monitor 8 runtimes spawned on other threads, then poll in the background from this one
+//! let mut monitor = RuntimeMonitor::new();
+//!
+//! for i in 0..8 {
+//!     let r = tokio::runtime::Builder::new_current_thread().enable_all().build().unwrap();
+//!     
+//!     monitor.add_runtime(None, Some(i), r.handle());
+//!
+//!     thread::spawn(move || r.block_on(async {
+//!         loop {
+//!             println!("Poll");
+//!             tokio::time::sleep(Duration::from_secs(1)).await;
+//!         }
+//!     }));
+//! }
+//!
+//! loop {
+//!     monitor.record_sample();
+//!
+//!     // record metrics roughly twice a second
+//!     std::thread::sleep(Duration::from_secs_f32(0.5));
+//! }
+//! ```
+
+use crate::telemetry::metrics::{metrics, Counter, Gauge};
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use tokio::runtime::{Handle, RuntimeMetrics};
+
+/// Monitors a set of runtimes, allowing metrics to be periodically collected for each tracked runtime.
+#[derive(Default)]
+pub struct RuntimeMonitor {
+    runtimes: Vec<RuntimeHandle>,
+}
+
+impl RuntimeMonitor {
+    /// Construct a new runtime monitor which initially monitors no runtimes.
+    pub fn new() -> Self {
+        Self {
+            runtimes: Vec::new(),
+        }
+    }
+
+    /// Add a runtime to the monitor, optionally with a name and/or id in case you are monitoring multiple runtimes.
+    ///
+    /// Runtimes should be uniquely identifiable by both label and id.
+    pub fn add_runtime(
+        &mut self,
+        runtime_name: Option<Arc<str>>,
+        runtime_id: Option<usize>,
+        handle: &Handle,
+    ) {
+        if self.runtimes.iter().any(|handle| {
+            (handle.runtime_name.is_some() != runtime_name.is_some())
+                || (handle.runtime_id.is_some() != runtime_id.is_some())
+        }) {
+            panic!("If you specify a runtime name or ID for one runtime, that setting must be specified for all");
```

Comment:
Don't panic, return `BootstrapError` instead

Reply:
I don't think this should be treated as a runtime error. If I could make this a compilation error without making the API painful, I would do so.

Reply:
I've made this an assertion.